File: solutions/haskell/bob/2/src/Bob.hs
module Bob (responseFor) where

import Data.Char(isUpper, isSpace, isLetter)

responseFor :: String -> String
responseFor xs
  | isSilent xs = "Fine. Be that way!"
  | isShouting xs && isQuestion xs = "Calm down, I know what I'm doing!"
  | isQuestion xs = "Sure."
  | isShouting xs = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    notSpace = filter (not. isSpace) xs
    letters = filter isLetter xs
    isShouting xs = any isLetter letters && all isUpper letters
    isQuestion xs = not (null notSpace) && last notSpace == '?'
    isSilent xs = null notSpace

File: solutions/r/bob/1/bob.R
bob <- function(input) {
  input <- trimws(input, which="both")
  if(input=="") {
    return("Fine. Be that way!")
  }
  length <- nchar(input)
  split <- unlist(strsplit(input,""))
  if(!grepl("[a-z]",input)&&grepl("[A-Z]",input)) {
    if(split[length]=="?") {
      return("Calm down, I know what I'm doing!")
    }
    return("Whoa, chill out!")
  }
  if(split[length]=="?") {
    return("Sure.")
  }
  return("Whatever.")
}
